tiledb/sm/filesystem/azure.cc

```diff
@@ -0,0 +1,716 @@
+          container_name, blob_path, offset, length, ss);
+  if (!request.valid()) {
+    return LOG_STATUS(Status::AzureError(
+        std::string("Read blob failed on: " + uri.to_string())));
+  }
+
+  azure::storage_lite::storage_outcome<void> outcome = request.get();
+  if (!outcome.success()) {
+    return LOG_STATUS(Status::AzureError(
+        std::string("Read blob failed on: " + uri.to_string())));
+  }
+
+  ss >> static_cast<char*>(buffer);
+
+  return Status::Ok();
+}
+
+Status Azure::remove_container(const URI& uri) const {
+  assert(client_);
+
+  // Empty container
+  RETURN_NOT_OK(empty_container(uri));
+
+  std::string container_name;
+  RETURN_NOT_OK(parse_azure_uri(uri, &container_name, nullptr));
+
+  std::future<azure::storage_lite::storage_outcome<void>> request =
+      client_->delete_container(container_name);
+  if (!request.valid()) {
+    return LOG_STATUS(Status::AzureError(
+        std::string("Remove container failed on: " + uri.to_string())));
+  }
+
+  azure::storage_lite::storage_outcome<void> outcome = request.get();
+  if (!outcome.success()) {
+    return LOG_STATUS(Status::AzureError(
+        std::string("Remove container failed on: " + uri.to_string())));
+  }
+
+  return wait_for_container_to_be_deleted(container_name);
+}
+
+Status Azure::remove_blob(const URI& uri) const {
+  assert(client_);
+
+  std::string container_name;
+  std::string blob_path;
+  RETURN_NOT_OK(parse_azure_uri(uri, &container_name, &blob_path));
+
+  std::future<azure::storage_lite::storage_outcome<void>> request =
+      client_->delete_blob(
+          container_name, blob_path, false /* delete_snapshots */);
+  if (!request.valid()) {
+    return LOG_STATUS(Status::AzureError(
+        std::string("Remove blob failed on: " + uri.to_string())));
+  }
+
+  azure::storage_lite::storage_outcome<void> outcome = request.get();
+  if (!outcome.success()) {
+    return LOG_STATUS(Status::AzureError(
+        std::string("Remove blob failed on: " + uri.to_string())));
+  }
+
+  return wait_for_blob_to_be_deleted(container_name, blob_path);
+}
+
+Status Azure::remove_dir(const URI& uri) const {
+  assert(client_);
+
+  std::vector<std::string> paths;
+  const URI uri_dir = uri.add_trailing_slash();
+  RETURN_NOT_OK(ls(uri_dir, &paths, ""));
+  for (const auto& path : paths) {
+    RETURN_NOT_OK(remove_blob(URI(path)));
+  }
+
+  return Status::Ok();
+}
+
+Status Azure::touch(const URI& uri) const {
+  assert(client_);
+
+  if (!uri.is_azure()) {
+    return LOG_STATUS(Status::AzureError(
+        std::string("URI is not an Azure URI: " + uri.to_string())));
+  }
+
+  std::string container_name;
+  std::string blob_path;
+  RETURN_NOT_OK(parse_azure_uri(uri, &container_name, &blob_path));
+
+  std::stringstream empty_ss;
+  std::vector<std::pair<std::string, std::string>> empty_metadata;
+  std::future<azure::storage_lite::storage_outcome<void>> request =
+      client_->upload_block_blob_from_stream(
```

Comment:
Noting that I believe this will overwrite the blob with an empty blob (same as the S3 backend, I think..). S3 and Azure both support "metadata" which can contain a (locally-generated) timestamp, but they don't otherwise seem to support a touch operation: https://stackoverflow.com/a/39695844/508431

Reply:
That's true. We have the same issue on S3. I'll open a clubhouse issue for this and handle them both at the same time.

Reply:
https://app.clubhouse.io/tiledb-inc/story/1588/provide-uri-normalization-for-azure-and-s3